File: apps/desktop/src-tauri/src/commands/sftp.rs
use serde::Serialize;
use tauri::ipc::Channel;
use tauri::State;

#[cfg(any(target_os = "android", target_os = "ios"))]
use crate::errors::LumaError;
use crate::errors::Result;
use crate::keystore::KeystoreState;
use crate::sftp::{
    self, DirectoryListing, SftpConnectResponse, SftpManager, SftpSessionInfo, TransferProgress,
    TransferStartResponse,
};
use crate::ssh;
use crate::AppState;

#[tauri::command]
pub async fn sftp_connect(
    state: State<'_, AppState>,
    manager: State<'_, SftpManager>,
    keystore_state: State<'_, KeystoreState>,
    host_id: String,
) -> Result<SftpConnectResponse> {
    manager
        .connect(&state.pool, &keystore_state, &host_id)
        .await
}

#[tauri::command]
pub async fn sftp_disconnect(
    manager: State<'_, SftpManager>,
    sftp_session_id: String,
) -> Result<()> {
    manager.disconnect(&sftp_session_id).await
}

#[tauri::command]
pub async fn sftp_sessions(manager: State<'_, SftpManager>) -> Result<Vec<SftpSessionInfo>> {
    Ok(manager.list())
}

#[tauri::command]
pub async fn sftp_list(
    manager: State<'_, SftpManager>,
    session_id: String,
    path: String,
) -> Result<DirectoryListing> {
    sftp::list(&manager, &session_id, &path).await
}

#[tauri::command]
pub async fn sftp_mkdir(
    manager: State<'_, SftpManager>,
    session_id: String,
    path: String,
) -> Result<()> {
    sftp::mkdir(&manager, &session_id, &path).await
}

#[tauri::command]
pub async fn sftp_rename(
    manager: State<'_, SftpManager>,
    session_id: String,
    from: String,
    to: String,
) -> Result<()> {
    sftp::rename(&manager, &session_id, &from, &to).await
}

#[tauri::command]
pub async fn sftp_delete(
    manager: State<'_, SftpManager>,
    session_id: String,
    path: String,
    recursive: bool,
) -> Result<()> {
    sftp::delete(&manager, &session_id, &path, recursive).await
}

#[tauri::command]
#[cfg(not(any(target_os = "android", target_os = "ios")))]
pub async fn local_list(path: Option<String>) -> Result<DirectoryListing> {
    sftp::local_list(path).await
}

#[tauri::command]
#[cfg(not(any(target_os = "android", target_os = "ios")))]
pub async fn local_mkdir(path: String) -> Result<()> {
    sftp::local_mkdir(path).await
}

#[tauri::command]
#[cfg(not(any(target_os = "android", target_os = "ios")))]
pub async fn local_rename(state: State<'_, AppState>, from: String, to: String) -> Result<()> {
    sftp::local_rename(from, to, state.app_data_dir.clone()).await
}

#[tauri::command]
#[cfg(not(any(target_os = "android", target_os = "ios")))]
pub async fn local_delete(state: State<'_, AppState>, path: String, recursive: bool) -> Result<()> {
    sftp::local_delete(path, recursive, state.app_data_dir.clone()).await
}

#[tauri::command]
pub async fn sftp_upload(
    manager: State<'_, SftpManager>,
    session_id: String,
    local_path: String,
    remote_path: String,
    on_progress: Channel<TransferProgress>,
) -> Result<TransferStartResponse> {
    sftp::sftp_upload(
        &manager,
        &session_id,
        &local_path,
        &remote_path,
        on_progress,
    )
    .await
}

#[tauri::command]
pub async fn sftp_download(
    state: State<'_, AppState>,
    manager: State<'_, SftpManager>,
    session_id: String,
    remote_path: String,
    local_path: String,
    on_progress: Channel<TransferProgress>,
) -> Result<TransferStartResponse> {
    sftp::sftp_download(
        &manager,
        &session_id,
        &remote_path,
        &local_path,
        &state.app_data_dir,
        on_progress,
    )
    .await
}

#[tauri::command]
pub async fn sftp_copy(
    manager: State<'_, SftpManager>,
    source_session_id: String,
    source_path: String,
    dest_session_id: String,
    dest_path: String,
    on_progress: Channel<TransferProgress>,
) -> Result<TransferStartResponse> {
    sftp::sftp_copy(
        &manager,
        &source_session_id,
        &source_path,
        &dest_session_id,
        &dest_path,
        on_progress,
    )
    .await
}

#[tauri::command]
pub async fn sftp_cancel(manager: State<'_, SftpManager>, transfer_id: String) -> Result<()> {
    manager.cancel_transfer(&transfer_id)
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalAttachUploadResponse {
    pub remote_path: String,
}

/// Upload a local file to a private staging directory on an SSH host so the
/// terminal UI can insert the returned remote path at the prompt.
///
/// Reuses an SFTP session the user already has open for the host — a second
/// authenticated connection per attachment costs a full handshake and can put
/// an interactive or 2FA prompt in the middle of attaching a file. Only a
/// session this command opened itself is closed afterwards.
#[tauri::command]
pub async fn terminal_attach_upload(
    state: State<'_, AppState>,
    keystore_state: State<'_, KeystoreState>,
    manager: State<'_, SftpManager>,
    host_id: String,
    local_path: String,
    file_name: Option<String>,
) -> Result<TerminalAttachUploadResponse> {
    ssh::validate_host_id(&host_id)?;
    // Reading the home directory doubles as the liveness check: a session that
    // has since died falls through to opening a dedicated one.
    let reusable = match manager.session_for_host(&host_id) {
        Some(session_id) => match manager.home_directory(&session_id).await {
            Ok(home) => Some((session_id, home)),
            Err(error) => {
                tracing::debug!(%error, "reusable SFTP session was unusable; opening a new one");
                None
            }
        },
        None => None,
    };
    let (session_id, home, opened_here) = match reusable {
        Some((session_id, home)) => (session_id, home, false),
        None => {
            let session = manager
                .connect(&state.pool, &keystore_state, &host_id)
                .await?;
            (session.sftp_session_id, session.initial_path, true)
        }
    };
    let upload_result = sftp::upload_attachment(
        &manager,
        &session_id,
        &home,
        &local_path,
        file_name.as_deref(),
    )
    .await;
    // Never close a session the user opened; only the one this command made.
    if opened_here {
        if let Err(error) = manager.disconnect(&session_id).await {
            tracing::warn!(%error, "failed to close SFTP session after attachment upload");
        }
    }
    upload_result.map(|remote_path| TerminalAttachUploadResponse { remote_path })
}

#[tauri::command]
pub async fn sftp_retry(
    manager: State<'_, SftpManager>,
    transfer_id: String,
    on_progress: Channel<TransferProgress>,
) -> Result<TransferStartResponse> {
    sftp::sftp_retry(&manager, &transfer_id, on_progress).await
}

/// Resolve the directory mobile downloads land in, creating it if needed.
///
/// iOS and Android have no folder picker (`tauri-plugin-dialog` returns
/// `FolderPickerNotImplemented` on both), so a folder download cannot ask the
/// user where to put it. It goes to the app's own Documents directory instead,
/// which `UIFileSharingEnabled` exposes in Files.app.
#[cfg(any(target_os = "android", target_os = "ios"))]
#[tauri::command]
pub async fn sftp_mobile_download_dir(app: tauri::AppHandle) -> Result<String> {
    use tauri::Manager;

    let dir = app
        .path()
        .document_dir()
        .map_err(|error| LumaError::InvalidInput(format!("no document directory: {error}")))?;
    tokio::fs::create_dir_all(&dir).await?;
    dir.to_str()
        .map(str::to_owned)
        .ok_or_else(|| LumaError::InvalidInput("document directory path is not valid UTF-8".into()))
}

/// Discard the empty placeholder an iOS save dialog leaves in Documents.
///
/// `saveFileDialog` cannot ask iOS for a destination path directly, so it
/// creates an empty file in the app's Documents directory and exports *that*,
/// returning the location the user picked. The placeholder is never cleaned up,
/// and `UIFileSharingEnabled` makes the strays visible in Files.app.
///
/// Only ever removes a zero-byte regular file directly inside Documents, and
/// never the path the save dialog returned: a finished download is not empty,
/// so a real file cannot be destroyed by a name collision.
#[cfg(any(target_os = "android", target_os = "ios"))]
#[tauri::command]
pub async fn sftp_discard_save_placeholder(
    app: tauri::AppHandle,
    file_name: String,
    saved_path: String,
) -> Result<()> {
    use tauri::Manager;

    // A bare file name only: anything with separators or traversal could point
    // outside Documents.
    if file_name.is_empty()
        || file_name.contains('/')
        || file_name.contains('\\')
        || file_name.contains('\0')
        || file_name == "."
        || file_name == ".."
    {
        return Err(LumaError::InvalidInput(
            "placeholder name must be a bare file name".into(),
        ));
    }

    let documents = app
        .path()
        .document_dir()
        .map_err(|error| LumaError::InvalidInput(format!("no document directory: {error}")))?;
    let placeholder = documents.join(&file_name);

    // The save dialog returns where the user actually put the file. If that is
    // the same path, there is no placeholder to discard -- only the real file.
    if let Some(saved) = crate::platform::picker_path(&saved_path) {
        if saved == placeholder {
            return Ok(());
        }
    }

    let metadata = match tokio::fs::symlink_metadata(&placeholder).await {
        Ok(metadata) => metadata,
        // Nothing there is the normal outcome on platforms that do not stage a
        // placeholder at all.
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => return Ok(()),
        Err(error) => return Err(error.into()),
    };
    if !metadata.is_file() || metadata.len() != 0 {
        return Ok(());
    }
    tokio::fs::remove_file(&placeholder).await?;
    Ok(())
}

File: apps/desktop/src/lib/sftp.ts
import { Channel, invoke } from "@tauri-apps/api/core";

/*
 * Typed invoke wrappers for the Phase 6 SFTP backend, mirroring the style of
 * src/lib/ssh.ts and src/lib/portForwards.ts. All types are camelCase; optional
 * fields arrive as `null`. The frontend has NO filesystem access of its own —
 * even "local" browsing goes through backend commands. Secrets are never
 * handled here.
 *
 * Transfers stream a Channel<TransferProgress>. After a transfer starts, runtime
 * errors arrive on that channel as the single terminal event (the invoke itself
 * resolves with the transferId). The backend forgets finished transfers, so the
 * store retains them for the visible queue.
 */

export type SftpKind = "file" | "dir" | "symlink" | "other";

export type SftpEntry = {
  name: string;
  /** Canonical absolute path, using the scope's separator style. */
  path: string;
  kind: SftpKind;
  size: number | null;
  /** Unix seconds, or null when the backend could not read it. */
  modifiedAt: number | null;
  /** Permission string (e.g. "rwxr-xr-x") on remote; often null locally. */
  permissions: string | null;
};

export type DirectoryListing = {
  /** Canonical path the backend resolved (may differ from the requested one). */
  path: string;
  entries: SftpEntry[];
};

export type TransferState =
  | "running"
  | "completed"
  | "failed"
  | "cancelled"
  /** Only used on per-entry ("entry") events for skipped symlinks. */
  | "skipped";

/**
 * Which facet of a directory transfer an event describes. Absent entirely for
 * single-file jobs, whose events carry only the original five fields (legacy).
 *  - "file": progress of the file currently being transferred, plus an
 *    aggregate snapshot of the whole job.
 *  - "aggregate": overall progress; `transferred`=bytesDone, `total`=totalBytes.
 *  - "entry": a per-entry outcome (skipped symlink or failed entry); `filePath`
 *    is the entry's path relative to the transfer root, always using "/".
 */
export type TransferProgressKind = "file" | "aggregate" | "entry";

/** Whole-job snapshot carried on directory "file" events. */
export type TransferAggregate = {
  totalBytes: number;
  bytesDone: number;
  totalFiles: number;
  filesDone: number;
  currentFilePath: string | null;
};

export type TransferProgress = {
  transferId: string;
  transferred: number;
  total: number | null;
  state: TransferState;
  errorMessage: string | null;
  /** Absent on single-file (legacy) events; set on every directory event. */
  progressKind?: TransferProgressKind;
  /** Relative "/"-separated path: current file on "file", entry on "entry". */
  filePath?: string;
  /** Whole-job snapshot; present on directory "file" events. */
  aggregate?: TransferAggregate;
  /** Byte offset a resumed file started at, present on the FIRST "file" (and
   * single-file running) event of a resumed transfer — `transferred` begins at
   * this offset. Absent for transfers that started from the beginning. */
  resumedFrom?: number;
};

export type SftpSessionInfo = { sftpSessionId: string; hostId: string };
export type SftpConnectResult = { sftpSessionId: string; initialPath: string };
export type TransferHandle = { transferId: string };

// Session lifecycle ----------------------------------------------------------

export function sftpConnect(hostId: string): Promise<SftpConnectResult> {
  return invoke<SftpConnectResult>("sftp_connect", { hostId });
}

export function sftpDisconnect(sftpSessionId: string): Promise<void> {
  return invoke<void>("sftp_disconnect", { sftpSessionId });
}

export function sftpSessions(): Promise<SftpSessionInfo[]> {
  return invoke<SftpSessionInfo[]>("sftp_sessions", {});
}

// Remote operations ----------------------------------------------------------

export function sftpList(sessionId: string, path: string): Promise<DirectoryListing> {
  return invoke<DirectoryListing>("sftp_list", { sessionId, path });
}

export function sftpMkdir(sessionId: string, path: string): Promise<void> {
  return invoke<void>("sftp_mkdir", { sessionId, path });
}

export function sftpRename(sessionId: string, from: string, to: string): Promise<void> {
  return invoke<void>("sftp_rename", { sessionId, from, to });
}

export function sftpDelete(
  sessionId: string,
  path: string,
  recursive: boolean,
): Promise<void> {
  return invoke<void>("sftp_delete", { sessionId, path, recursive });
}

// Local operations (backend-mediated; the frontend has no fs access) ---------

export function localList(path: string | null): Promise<DirectoryListing> {
  return invoke<DirectoryListing>("local_list", { path });
}

export function localMkdir(path: string): Promise<void> {
  return invoke<void>("local_mkdir", { path });
}

export function localRename(from: string, to: string): Promise<void> {
  return invoke<void>("local_rename", { from, to });
}

export function localDelete(path: string, recursive: boolean): Promise<void> {
  return invoke<void>("local_delete", { path, recursive });
}

/**
 * The directory mobile downloads land in. Mobile has no folder picker, so a
 * folder download goes to the app's own Documents directory, which is visible
 * in Files.app. Mobile only — the command is not registered on desktop.
 */
export function sftpMobileDownloadDir(): Promise<string> {
  return invoke<string>("sftp_mobile_download_dir");
}

/**
 * Discard the empty placeholder an iOS save dialog stages in Documents.
 * `saveFileDialog` has no way to ask iOS for a path, so it exports a throwaway
 * empty file and returns where the user put the real one; the placeholder is
 * left behind. No-op when nothing was staged. Mobile only.
 */
export function sftpDiscardSavePlaceholder(
  fileName: string,
  savedPath: string,
): Promise<void> {
  return invoke<void>("sftp_discard_save_placeholder", { fileName, savedPath });
}

// Transfers ------------------------------------------------------------------

export function sftpUpload(
  sessionId: string,
  localPath: string,
  remotePath: string,
  onProgress: (progress: TransferProgress) => void,
): Promise<TransferHandle> {
  const channel = new Channel<TransferProgress>();
  channel.onmessage = onProgress;
  return invoke<TransferHandle>("sftp_upload", {
    sessionId,
    localPath,
    remotePath,
    onProgress: channel,
  });
}

export function sftpDownload(
  sessionId: string,
  remotePath: string,
  localPath: string,
  onProgress: (progress: TransferProgress) => void,
): Promise<TransferHandle> {
  const channel = new Channel<TransferProgress>();
  channel.onmessage = onProgress;
  return invoke<TransferHandle>("sftp_download", {
    sessionId,
    remotePath,
    localPath,
    onProgress: channel,
  });
}

/**
 * Copy a file or directory directly between two connected hosts. The bytes
 * stream through the app; nothing is written to the local filesystem. The two
 * sessions may belong to the same host.
 */
export function sftpCopy(
  sourceSessionId: string,
  sourcePath: string,
  destSessionId: string,
  destPath: string,
  onProgress: (progress: TransferProgress) => void,
): Promise<TransferHandle> {
  const channel = new Channel<TransferProgress>();
  channel.onmessage = onProgress;
  return invoke<TransferHandle>("sftp_copy", {
    sourceSessionId,
    sourcePath,
    destSessionId,
    destPath,
    onProgress: channel,
  });
}

export type AttachUploadResult = { remotePath: string };

/**
 * Upload a local file to the host's private staging directory
 * (~/.luma/attachments) over a short-lived SFTP session, returning the
 * absolute remote path. Used by the terminal "Attach file" action; unlike
 * sftpUpload this resolves only after the transfer has fully completed.
 */
export function terminalAttachUpload(
  hostId: string,
  localPath: string,
  fileName?: string,
): Promise<AttachUploadResult> {
  return invoke<AttachUploadResult>("terminal_attach_upload", {
    hostId,
    localPath,
    fileName: fileName ?? null,
  });
}

export function sftpCancel(transferId: string): Promise<void> {
  return invoke<void>("sftp_cancel", { transferId });
}

export function sftpForgetTransfers(transferIds: string[]): Promise<void> {
  return invoke<void>("sftp_forget_transfers", { transferIds });
}

/**
 * Retry the failed / incomplete entries of a finished transfer. Returns a NEW
 * transferId that owns the retry — the caller must rebind its queue row to it
 * for any subsequent cancel or progress handling. Rejects with invalid-input
 * "unknown transfer", "transfer is still running", or "transfer has no failed
 * or incomplete entries to retry".
 */
export function sftpRetry(
  transferId: string,
  onProgress: (progress: TransferProgress) => void,
): Promise<TransferHandle> {
  const channel = new Channel<TransferProgress>();
  channel.onmessage = onProgress;
  return invoke<TransferHandle>("sftp_retry", {
    transferId,
    onProgress: channel,
  });
}

// Formatting + path helpers --------------------------------------------------

const BYTE_UNITS = ["KiB", "MiB", "GiB", "TiB", "PiB"];

/** Human-readable byte size. Defensive against null / negative / non-finite. */
export function formatBytes(bytes: number | null | undefined): string {
  if (bytes == null || !Number.isFinite(bytes) || bytes < 0) return "—";
  if (bytes < 1024) return `${bytes} B`;
  let value = bytes / 1024;
  let unit = 0;
  while (value >= 1024 && unit < BYTE_UNITS.length - 1) {
    value /= 1024;
    unit += 1;
  }
  return `${value.toFixed(value >= 100 ? 0 : 1)} ${BYTE_UNITS[unit]}`;
}

/** Format a transfer rate in bytes/second. */
export function formatRate(bytesPerSecond: number): string {
  if (!Number.isFinite(bytesPerSecond) || bytesPerSecond <= 0) return "";
  return `${formatBytes(bytesPerSecond)}/s`;
}

/** Format a unix-seconds timestamp defensively (null / bogus -> em dash). */
export function formatModified(unixSeconds: number | null | undefined): string {
  if (unixSeconds == null || !Number.isFinite(unixSeconds) || unixSeconds <= 0) {
    return "—";
  }
  const date = new Date(unixSeconds * 1000);
  if (Number.isNaN(date.getTime())) return "—";
  const now = Date.now();
  const sameYear = date.getFullYear() === new Date(now).getFullYear();
  return date.toLocaleString(undefined, {
    year: sameYear ? undefined : "numeric",
    month: "short",
    day: "numeric",
    hour: "2-digit",
    minute: "2-digit",
  });
}

/** Passthrough for the backend permission string (null -> em dash). */
export function formatPermissions(permissions: string | null | undefined): string {
  return permissions && permissions.length > 0 ? permissions : "—";
}

/**
 * Infer the separator style of a canonical path. Remote paths are always "/";
 * local paths use whatever the backend's canonical form returned (Windows uses
 * "\"). We never assume "/" for local paths.
 */
export function inferSeparator(path: string): "/" | "\\" {
  return path.includes("\\") ? "\\" : "/";
}

/** Join a directory and a child name using the given separator. */
export function joinPath(base: string, name: string, separator: "/" | "\\"): string {
  if (base.endsWith(separator)) return `${base}${name}`;
  return `${base}${separator}${name}`;
}

/** Join within the remote scope (always "/"). */
export function remoteJoin(base: string, name: string): string {
  return joinPath(base, name, "/");
}

/**
 * Parent of a canonical path, or null when already at a root. Derived purely by
 * string operations on the canonical path so it works for both "/" (remote /
 * unix) and "\" (Windows) styles, including drive roots like "C:\".
 */
export function parentPath(path: string, separator: "/" | "\\"): string | null {
  const trimmed = path.replace(/[/\\]+$/, "");
  const index = trimmed.lastIndexOf(separator);
  if (index < 0) return null;
  if (separator === "/") {
    return index === 0 ? "/" : trimmed.slice(0, index);
  }
  const parent = trimmed.slice(0, index);
  if (parent === "") return null;
  // Keep the trailing separator for a drive root ("C:" -> "C:\").
  if (/^[A-Za-z]:$/.test(parent)) return `${parent}${separator}`;
  return parent;
}

/**
 * Split a canonical path into clickable breadcrumb segments. Each segment
 * carries the absolute path to navigate to when clicked.
 */
export function breadcrumbSegments(
  path: string,
  separator: "/" | "\\",
): { label: string; path: string }[] {
  const segments: { label: string; path: string }[] = [];
  if (separator === "/") {
    segments.push({ label: "/", path: "/" });
    const parts = path.split("/").filter(Boolean);
    let acc = "";
    for (const part of parts) {
      acc = `${acc}/${part}`;
      segments.push({ label: part, path: acc });
    }
    return segments;
  }
  // Windows-style: first part is the drive ("C:").
  const parts = path.split("\\").filter(Boolean);
  let acc = "";
  parts.forEach((part, i) => {
    if (i === 0) {
      acc = `${part}\\`;
      segments.push({ label: part, path: acc });
    } else {
      acc = acc.endsWith("\\") ? `${acc}${part}` : `${acc}\\${part}`;
      segments.push({ label: part, path: acc });
    }
  });
  if (segments.length === 0) segments.push({ label: path, path });
  return segments;
}
